### ISSUE TITLE
Kotlin + Play

### ISSUE THREAD

**Issue by raderio** [playframework/playframework#7243]
Is it possible to use Kotlin with Play?

**Comment by shigengtian** [comment]
Why do you want to use Kotlin?
I think Scala and java is enough.

**Comment by raderio** [comment]
I don't like implicits in Scala, nullabilty in Kotlin is better, ...
I don't want to make a holywar, I like Play and I like Kotlin, so I want to use them both together.

**Comment by domdorn** [comment]
hi @raderio 
At the moment, we don't provide native kotlin "support". In theory, it should work if you're using 
https://github.com/pfn/kotlin-plugin
and the java-apis. You'll probably run into quite a few issues. 
The question is: "what are you looking for" ? If you just want a working framework, the play + kotlin combo is the wrong approach at the moment. If you're willing to spend time working around issues, you might give it a try.

**Comment by wsargent** [comment]
If you use the Java APIs with Kotlin, you will probably be okay since it'll all produce the same bytecode in the end.  However, there's no plan to produce a Kotlin API.

**Comment by oexza** [comment]
There really is no need for a kotlin api, all that is needed is a
decent integration between kotlin and sbt. That should work.

**Comment by adrianpowell** [comment]
I've tried going down the Kotlin + Play path for a ways using https://github.com/pfn/kotlin-plugin

Neither Kotlin nor Scala have support (or planned support) for a mixed-mode Java+Kotlin+Scala compiler so you need to pick one or two, but not three. Since Twirl templates are first compiled to a .scala file, this means that you can't use Kotlin and Twirl at the same time either.

**Comment by laguiz** [comment]
It would be very interesting to see lightbend getting closer to kotlin community :-)

**Comment by richdougherty** [comment]
> Neither Kotlin nor Scala have support (or planned support) for a mixed-mode Java+Kotlin+Scala compiler so you need to pick one or two, but not three. Since Twirl templates are first compiled to a .scala file, this means that you can't use Kotlin and Twirl at the same time either.

It would be great if _javac_ was pluggable so that it could support multiple source file types…

**Comment by gmethvin** [comment]
I'm reopening this because it's clear some work is required here. In particular:
 - Make sure our APIs have proper NotNull/Nullable annotations in Java so they work well in Kotlin. This will also be useful to our Java users.
 - Provide a recommended way to set up a Play Kotlin/Java build. This could be done with Gradle or Maven, and there are already plugins supporting those build systems. We might need to suggest using an alternate templating system that doesn't use Scala.

**Comment by domdorn** [comment]
hmm... if you separate your app into different submodules, you might get it to work without much changing? 
so one module for templates/views, one for view-model classes (and possible "view interfaces" and "route interfaces") and one for the main application. 

first compile the view-model classes module.. then compile the templates/views module, in the end bind everything together (e.g. using dependency injection) in the main app... 
might be tricky, but might work.. haven't tried it thought. 

If we could get interfaces/traits for views & routes, it might be easier to make those "pluggable", what do you think @gmethvin ?

**Comment by laguiz** [comment]
@gmethvin
My most common scenario is Play Java without templating, only rest/json for SPA application.
A quick POC project would be very appreciated for this simpler scenario.

**Comment by gmethvin** [comment]
@laguiz Then does https://github.com/pfn/kotlin-plugin work for you? That should be all you need. The only issue is https://github.com/pfn/kotlin-plugin/issues/15.

If it does work for you, it would be good to get some feedback on any API improvements we could make to improve the experience for kotlin users.

**Comment by raderio** [comment]
https://github.com/67726e/gradle-twirl

**Comment by raderio** [comment]
https://github.com/playframework/twirl/issues/156

**Comment by tonilopezmr** [comment]
Please fix it, you are losing all kotlin community going to spring.

**Comment by gmethvin** [comment]
I think what needs to be defined is what "fix it" means here. It's already possible to write Play code in Kotlin: https://github.com/gmethvin/play-kotlin-rest-api-example.

If you want an alternative to twirl, there is some discussion on Kotlin templating here: https://discuss.kotlinlang.org/t/template-engine/866/16. You can also look at rocker: https://github.com/fizzed/rocker

So it looks like we need:
 - an official example, showing how to do templates, maybe using gradle instead of sbt
 - a review of our Java APIs for null-safety and other Kotlin usability issues
 - documentation on how to do basic things in Kotlin using the Play Java API

I think all these things should be doable for someone familiar with Kotlin. I've marked this as "help wanted" since it would be best for someone from the Kotlin community to take this on.

**Comment by tonilopezmr** [comment]
@gmethvin Thanks for your answer, yes I have created an example like `play-kotlin-rest-api-example` and It's works with some limitations like running or testing automatically. I feel that Play Framework maintainers don't want to adopt kotlin while spring framework is supporting kotlin, I think.

**Comment by gmethvin** [comment]
> I feel that Play Framework maintainers don't want to adopt kotlin while spring framework is supporting kotlin, I think.

We are completely open to contributions to improve our Kotlin support. We just have a smaller team and plenty of other things going on, so this is not the priority for the maintainers at the moment. This is why we're looking for contributions from the community.

**Comment by wsargent** [comment]
> I feel that Play Framework maintainers don't want to adopt kotlin while spring framework is supporting kotlin

It's one of the odd parts of the job -- no matter how much work you put in, there's always more to do and everyone feels like they're shortchanged on time and attention.  There's even a whole series of blog posts on what it feels like to be an OSS maintainer:

https://nolanlawson.com/2017/03/05/what-it-feels-like-to-be-an-open-source-maintainer/

**Comment by tonilopezmr** [comment]
@gmethvin oh! I didn't know that! 

 @wsargent yes I know how much work you put in but I have seen comments about play framework and kotlin since 2 years ago but in the recent months kotlin community has grown a lot. I will look over how I can help you with the kotlin community.

Thanks for your answers!

**Comment by ibinti** [comment]
It is nice to see some interests in Play + Kotlin.
I've been using sbt Play Scala for a while and then I came across Kotlin.
I wanted to use Play Scala with Kotlin. My use case is rather simple.
And my work around is to use Gradle + Play Scala + Kotlin.
Play Scala handles interfacing with routing and action controller, and everything else is done in Kotlin.

I hope Play will support Kotlin better in the future.

**Comment by gmethvin** [comment]
If you're using Kotlin with Play, it's going to be best to use the Java API. Scala uses completely different data types than Java, whereas Kotlin is designed to integrate more directly with Java types.

It's already possible to do integrate Play with Kotlin a bunch of different ways. If you want you could try using something like https://github.com/gmethvin/play-java-functional-actions, which would be less verbose in Kotlin than it is in Java and more closely matches the style used in Play Scala actions.

**Comment by tonilopezmr** [comment]
I'm using Play framework Java API with sbt modifying the actual [kotlin-plugin](https://github.com/pfn/kotlin-plugin)

**Comment by mkurz** [comment]
@tonilopezmr If you could do that I think that could be helpful for others. Will be interesting what your approach is. Thank you!

**Comment by zaxebo1** [comment]
so as already requested in https://github.com/playframework/playframework/issues/7243#issuecomment-295375849 ,  **whether sbt and kotlin are integrated now?**

**Comment by tonilopezmr** [comment]
Hey @mkurz @gmethvin, after 6 months into it, I have written a [blog post talking about it](https://blog.karumi.com/our-adventure-using-play-framework-in-kotlin/ ) and we have two big projects in production using it, and this is the project example [Play framework in kotlin](https://github.com/Karumi/play-framework-kotlin) 

Thanks guys!

**Comment by mkurz** [comment]
@tonilopezmr Thanks I will check it out in detail if I have more time. I just scrolled through the blog post, but you didn't mention templating it seems - e.g. what about templates with twirl (which actually result into Scala files)? Did you use another templating system?

**Comment by tonilopezmr** [comment]
Yes, we don't work with SPA and we don't investigate about it, but I tried to do a hello world at the beginning and for twirl, you can't use kotlin just Scala or Java.

**Comment by HugoTigre** [comment]
I've been working with Play for about 4 years, I like Java, Scala and Kotlin, that said, Kotlin has gain in popularity mainly because it has a smaller learning curve than Scala, more compatible with java and of course Android.

I think Play doesn't need to support all 3 languages at once, but maybe just Java/Scala or Java/Kotlin, don't know if that makes sense. Kotlin is being supported everywhere, by not supporting Kotlin, Play will loose a lot of people and that would be sad. If Play would to support Kotlin I would use it. Has it is right now I wouldn't use it in production with Kotlin until officially supported.

**Comment by damien-roche** [comment]
Over the last 12 months I went from 6 years of Ruby to Elixir and Go. I would jump to Kotlin and Play in a heartbeat if Play had the support. I understand the interop with Java means it makes sense to start with Play and Java, but it would be a major draw if the messy integration work was already done.

I tried Play many years ago with Scala and enjoyed it, but if I'm tackling the JVM them I'm going to pick a relatively simple language to hold my hands. Java has too many warts and isn't compelling enough over what I already have. Kotlin has my attention. Therefore, Kotlin + Play = :heart_eyes:

**Comment by raderio** [comment]
@damien-roche just use https://github.com/micronaut-projects/micronaut-core + https://github.com/fizzed/rocker

**Comment by damien-roche** [comment]
Appreciate the suggestion, but that's a micro framework. I have use for a full-featured MVC framework, not micro-services. I may use `micronaut` for something else, though, so thanks for the ref.

**Comment by raderio** [comment]
MIcronaut is just like Spring Boot. It was inspired by Spring Boot, but has compile time dependency injection.
